From 179298daa1e9842a58b89a5b638fd2de54e7eae6 Mon Sep 17 00:00:00 2001
From: Alex Williamson <alex.williamson@redhat.com>
Date: Fri, 1 Aug 2014 18:12:56 -0500
Subject: [CHANGE 5/5] Ignore PCI autoboot device location if set to 00:00.0
To: rhvirt-patches@redhat.com,
    jen@redhat.com

Message-id: <20140513195643.6596.14094.stgit@bling.home>
Patchwork-id: 58836
O-Subject: [RHEL6.6 gpxe PATCH 5/5] [prefix] Ignore PCI autoboot device location if set to 00:00.0
Bugzilla: 1057249
RH-Acked-by: Jeffrey Cody <jcody@redhat.com>
RH-Acked-by: Stefan Hajnoczi <stefanha@redhat.com>
RH-Acked-by: Laszlo Ersek <lersek@redhat.com>

From: Michael Brown <mcb30@ipxe.org>

Bugzilla: 1057249
Upstream commit: ff1e7fc72b138fe66de1452a311b7e871f33e201 (ipxe)

qemu can load an option ROM which is not associated with a particular
PCI device using the "-option-rom" syntax.  Under these circumstances,
we should ignore the PCI bus:dev.fn address that we expect to find in
%ax on entry to the initialisation vector.

Fix by using the PCI bus:dev.fn address only if it is non-zero.  Since
00:00.0 will always be the host bridge, it can never be the address of
a network card.

Reported-by: Alex Williamson <alex.williamson@redhat.com>
Signed-off-by: Michael Brown <mcb30@ipxe.org>
Signed-off-by: jen <jen@redhat.com>
---
 src/arch/i386/core/pci_autoboot.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/src/arch/i386/core/pci_autoboot.c b/src/arch/i386/core/pci_autoboot.c
index 9d7a1e9..2799555 100644
--- a/src/arch/i386/core/pci_autoboot.c
+++ b/src/arch/i386/core/pci_autoboot.c
@@ -32,8 +32,10 @@ uint16_t __bss16 ( autoboot_busdevfn );
  */
 static void pci_autoboot_init ( void ) {
 
-	autoboot_device.bus_type = BUS_TYPE_PCI;
-	autoboot_device.location = autoboot_busdevfn;
+	if ( autoboot_busdevfn ) {
+		autoboot_device.bus_type = BUS_TYPE_PCI;
+		autoboot_device.location = autoboot_busdevfn;
+	}
 }
 
 /** PCI autoboot device initialisation function */
-- 
1.9.3

